src/charm.py

```diff
@@ -574,20 +575,25 @@ def _on_update_status(self, event: UpdateStatusEvent):
         if self.unit.is_leader():
             self._handle_reconfigure(event)
 
-        # update the units status based on it's replica set config and backup status. An error in
-        # the status of MongoDB takes precedence over pbm status.
+        # retrieve statuses of different services running on Charmed MongoDB
         mongodb_status = build_unit_status(self.mongodb_config, self._unit_ip(self.unit))
-        pbm_status = self.backups.get_pbm_status()
-        if (
-            not isinstance(mongodb_status, ActiveStatus)
-            or not self.model.get_relation(
-                S3_RELATION
-            )  # if s3 relation doesn't exist only report MongoDB status
-            or isinstance(pbm_status, ActiveStatus)  # pbm is ready then report the MongoDB status
-        ):
-            self.unit.status = mongodb_status
-        else:
-            self.unit.status = pbm_status
+        shard_status = self.shard.get_shard_status() if self.is_role(Config.Role.SHARD) else None
+        config_server_status = (
+            self.config_server.get_config_server_status()
+            if self.is_role(Config.Role.CONFIG_SERVER)
+            else None
+        )
+        pbm_status = (
+            self.backups.get_pbm_status() if self.model.get_relation(S3_RELATION) else None
+        )
```

Comment:
would it make sense to put all these in a `get_status()` or `compute_unit_status()` method? where the priorisation also happens?

Reply:
Great idea!